### PR TITLE
Mpuncel/old rc node transfer

### PR DESCRIPTION
**Pull request by mpuncel**

### CODE REVIEW

File: pkg/roll/integration_test.go [hunk 1]
```diff
@@ -113,6 +120,72 @@ func TestCleanupOldRCHappy(t *testing.T) {
...
+	switch err {
+	case rcstore.NoReplicationController:
+		t.Fatal("old RC was deleted even though it wasn't zeroed out")
+	case nil:
+	default:
+		t.Fatalf("unexpected error when checking that old RC was deleted: %s", err)
+	}
+}
+
 // errorOnceChannelAlerter returns an error the first time you call Alert() and
```
Comment (mpuncel) [inline pkg/roll/integration_test.go]: i tried writing a test that an RC doesn't get deleted if it has a 0 replica count but still has current pods. unfortunately i'm not sure how to write that test without restructuring the code, basically i want to prove that the `RetryOrQuit()` call runs at least once and didn't add anything to the transaction? Both of which are not inspectable things. I think it's okay

File: pkg/rc/replication_controller_test.go [hunk 2]
```diff
@@ -1425,6 +1425,266 @@ func TestNodeTransferFailsOnScheduleStep(t *testing.T) {
...
+	current := make(types.PodLocations, 0)
+
+	eligible := []types.NodeName{"node1", "node2", "node3", "node4", "node5"}
+
+	err := rc.addPods(rcFields, current, eligible)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// now confirm that 5 pods were scheduled
```
Comment (spenceral) [inline pkg/rc/replication_controller_test.go]: Looks like a copy pasted comment from the previous test. Not correct here

**Comment by mpuncel** [comment]
I think there are still some edge cases to sort out here

**Comment by mpuncel** [comment]
OK i believe this to be complete now. will do some testing.

High level changes:

1) disabled RCs now perform node transfers
2) if the scheduleWithSession() step encounters a transaction conflict, it continues to waiting for healthy. Rationale is that the exact manifest doesn't matter, it just needs the new node to be healthy
3) rolling updates exclude ineligible nodes from health count for dynamic RCs (because an ineligible node can be unscheduled at any moment)

File: pkg/rc/replication_controller.go [hunk 3]
```diff
@@ -443,8 +446,68 @@ func (rc *replicationController) removePods(rcFields fields.RC, current types.Po
...
+				// (because it will rely on a different enabled RC to
+				// do it implicitly)
+				//
+				// One final safety restriction we put in place
+				// is that a disabled RC should only unschedule
+				// ineligible nodes if its replica count is
+				// zero, which we expect to occur at the end of
+				// a rolling update. This protection is
+				// necessary to prevent the rolling update from
+				// counting this ineligible (presumably
```
Comment (rudle) [inline pkg/rc/replication_controller.go]: I don't agree that ineligible nodes will be healthy in all or even most cases. I don't think it's material for the logic tho.

File: pkg/rc/replication_controller.go [hunk 4]
```diff
@@ -443,8 +446,68 @@ func (rc *replicationController) removePods(rcFields fields.RC, current types.Po
...
+				// after the replica count change is executed,
+				// cluster might fall below the minimum health
+				// requirement of the rolling update.
+				//
+				// If the RCs replica count is zero, we can
+				// reliably assume that the rolling update is
+				// over so it is always safe to unschedule the
+				// old node. This is enforced at the beginning
+				// of the removePods() function in the first
+				// return
```
Comment (spenceral) [inline pkg/rc/replication_controller.go]: WDYT about breaking up the comment and putting smaller comments closer to where the relevant code is? I had to reread it and think to determine what part of this comment is executed by the `if disabled: break`. It might be simpler to have it only say something like "Don't unschedule eligible nodes since that is implicitly performed by the new RC during roll when this RC is disabled" here,  add a small, relevant comment above `if !ok`, etc..

File: pkg/roll/run_update.go [hunk 5]
```diff
@@ -684,7 +713,31 @@ func (u *update) countHealthy(id rcf.ID, checks map[types.NodeName]health.Result
 		ret.Unknown = ret.Desired - ret.Current
 	}
 
+	var eligibleNodes []types.NodeName
+	if rcFields.AllocationStrategy == rcf.DynamicStrategy {
+		fmt.Println("calling bro")
```
Comment (rudle) [inline pkg/roll/run_update.go]: 🕶

File: pkg/rc/replication_controller_test.go [hunk 6]
```diff
@@ -1425,6 +1428,266 @@ func TestNodeTransferFailsOnScheduleStep(t *testing.T) {
...
+
+	// Set the count to zero, and now we should see the ineligible nodes become unscheduled
+	rcFields.ReplicasDesired = 0
+
+	err = rc.removePods(rcFields, currentPods, eligible)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// now confirm that 5 pods were scheduled
```
Comment (spenceral) [inline pkg/rc/replication_controller_test.go]: incorrect comment. Test confirms that 3 pods were scheduled

File: pkg/rc/replication_controller.go [hunk 7]
```diff
@@ -255,6 +243,7 @@ func (rc *replicationController) meetDesires(rcFields fields.RC) error {
 	nodesChanged := false
 	switch {
 	case rcFields.ReplicasDesired > len(current):
+		rc.nodeTransferMu.Lock()
 		if rc.nodeTransfer.quit != nil {
```
[the remark this answers is not in the document]
Comment (mpuncel) [reply pkg/rc/replication_controller.go]: we unconditionally reach the unlock on 259 so it's fine. I also want to protect the `rc.nodeTransfer.newNode` accesses

File: pkg/rc/replication_controller.go [hunk 8]
```diff
@@ -410,12 +404,23 @@ func (rc *replicationController) alertInfo(rcFields fields.RC, msg string) alert
...
 
 func (rc *replicationController) removePods(rcFields fields.RC, current types.PodLocations, eligible []types.NodeName) error {
+	// If the RC is disabled, we don't want to remove any pods with one
+	// exception: if the ReplicasDesired count is zero and their are "ineligible"
+	// nodes. In this case, we want to unschedule all ineligible nodes.
+	//
+	// An ineligible node is defined as a node that appears in "current" but not
+	// in "eligible". This means that the scheduler has decided this node is no
+	// longer eligible.
+	if rcFields.ReplicasDesired != 0 && rcFields.Disabled {
```
Comment (mpuncel) [inline pkg/rc/replication_controller.go]: hmm actually i think we want to just rely on the transfer finishing instead of having this now

File: pkg/roll/run_update.go [hunk 9]
```diff
@@ -684,7 +713,31 @@ func (u *update) countHealthy(id rcf.ID, checks map[types.NodeName]health.Result
...
+		eligibleNodes, err = u.scheduler.EligibleNodes(rcFields.Manifest, rcFields.NodeSelector)
+		if err != nil {
+			return ret, util.Errorf("could not enumerate eligible nodes for %s: %s", rcFields.ID, err)
+		}
+	}
+
 	for _, pod := range currentPods {
+		if rcFields.AllocationStrategy == rcf.DynamicStrategy {
```
Comment (spenceral) [inline pkg/roll/run_update.go]: nit: you could instead check for AllocationStrategy != Dynamic and continue to reduce the indentation. Kind of in the style of https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: pkg/roll/run_update.go [hunk 10]
```diff
@@ -684,7 +713,31 @@ func (u *update) countHealthy(id rcf.ID, checks map[types.NodeName]health.Result
...
+			nodeEligible := false
+			for _, eligibleNode := range eligibleNodes {
+				if eligibleNode == pod.Node {
+					nodeEligible = true
+					break
+				}
+			}
+			if !nodeEligible {
+				continue
+			}
```
Comment (spenceral) [inline pkg/roll/run_update.go]: I can't tell what this if block does. Doesn't the code go the next iteration of the loop anyway?

Comment (mpuncel) [reply pkg/roll/run_update.go]: i'll add a comment to clarify. Basically I don't want to increment `ret.Healthy` or `ret.Real` if the node isn't eligible

Comment (spenceral) [reply pkg/roll/run_update.go]: Nevermind. I totally misinterpreted the truncated file on github and thought the loop ended 🙃

File: pkg/roll/run_update.go (hunk 9 above)
Comment (mpuncel) [reply pkg/roll/run_update.go]: I don't think that would reduce indentation and it would make the if block longer since I would have to include the section that counts Healthy

Comment (spenceral) [reply pkg/roll/run_update.go]: You're right. This is the same loop I misread the end of. Ignore me

**Comment by mpuncel** [comment]
getting a test timeout in travis and not locally 🤔
